test/integration/client/client_test.go

```diff
@@ -795,3 +801,58 @@ func TestSelfLinkOnNamespace(t *testing.T) {
 
 	runSelfLinkTestOnNamespace(t, c, "default")
 }
+
+func TestEventCompatibility(t *testing.T) {
+	result := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--disable-admission-plugins", "ServiceAccount"}, framework.SharedEtcd())
+	defer result.TearDownFn()
+
+	client := clientset.NewForConfigOrDie(result.ClientConfig)
+
+	testPod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			SelfLink:  "/api/version/pods/foo",
+			Name:      "foo",
+			Namespace: "baz",
+			UID:       "bar",
+		},
+	}
+
+	regarding, err := ref.GetPartialReference(scheme.Scheme, testPod, ".spec.containers[1]")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	related, err := ref.GetPartialReference(scheme.Scheme, testPod, ".spec.containers[0]")
+	if err != nil {
+		t.Fatal(err)
+	}
+	stopCh := make(chan struct{})
+	oldBroadcaster := record.NewBroadcaster()
+	oldRecorder := oldBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "integration"})
+	oldBroadcaster.StartRecordingToSink(&typedv1.EventSinkImpl{Interface: client.CoreV1().Events("")})
+	oldRecorder.Eventf(regarding, v1.EventTypeNormal, "started", "note")
+
+	newBroadcaster := events.NewBroadcaster(&events.EventSinkImpl{Interface: client.EventsV1beta1().Events("")})
+	newRecorder := newBroadcaster.NewRecorder(scheme.Scheme, "k8s.io/kube-scheduler")
+	newBroadcaster.StartRecordingToSink(stopCh)
+	newRecorder.Eventf(regarding, related, v1.EventTypeNormal, "memoryPressure", "killed", "memory pressure")
+
+	v1beta1Events, err := client.EventsV1beta1().Events("").List(metav1.ListOptions{})
```

Comment:
Also as discussed offline - Eventf() is asynchronous, which introduces a race here. We should probably retry with timeout here.